wazo_auth/services/saml.py

```diff
@@ -0,0 +1,146 @@
+            # TODO(pc-m): Check that the configured domains exists for a tenant
+            raw_saml_config['relay_state'] = domain
+            raw_saml_config.update(global_saml_config)
+            try:
+                saml_config = SAMLConfig()
+                saml_config.load(raw_saml_config)
+                saml_client = Saml2Client(config=saml_config)
+                logger.debug('SAML config : %s', vars(saml_config))
+                self._saml_clients[domain] = saml_client
+            except Exception:
+                logger.exception('Error during SAML client init for domain %s', domain)
+
+    def get_client(self, domain):
+        return self._saml_clients[domain]
+
+    def prepare_redirect_response(
+        self,
+        redirect_url,
+        domain,
+    ):
+        saml_session_id = secrets.token_urlsafe(16)
+        client = self.get_client(domain)
+        reqid, info = client.prepare_for_authenticate(relay_state=domain)
+
+        self._outstanding_requests[reqid] = SamlAuthContext(
+            saml_session_id,
+            redirect_url,
+            domain,
+        )
+        location = [i for i in info['headers'] if i[0] == 'Location'][0][1]
+        return location, saml_session_id
+
+    def process_auth_response(self, url, remote_addr, form_data):
+        saml_client = self.get_client(form_data.get('RelayState'))
+        conv_info = {
+            "remote_addr": remote_addr,
+            "request_uri": url,
+            "entity_id": saml_client.config.entityid,
+            "endpoints": saml_client.config.getattr("endpoints", "sp"),
+        }
+
+        response: None = saml_client.parse_authn_request_response(
+            form_data['SAMLResponse'],
+            BINDING_HTTP_POST,
+            self._outstanding_requests,
+            None,
+            conv_info=conv_info,
+        )
+
+        logger.debug('SAML SP response: %s', response)
+        logger.info('SAML response AVA: %s', response.ava)
+
+        session_data: Optional[SamlAuthContext] = self._outstanding_requests.get(
+            response.session_id()
+        )
+        if session_data:
+            update = {'response': response, 'login': response.ava['name']}
+            self._outstanding_requests[response.session_id()] = replace(
+                session_data, **update
+            )
+            return session_data.redirect_url
+        else:
+            return None
+
+    def get_user_login_and_remove_context(self, saml_session_id):
+        logger.debug('sessions %s', self._outstanding_requests)
+        for key in self._outstanding_requests:
+            if self._outstanding_requests[key].saml_session_id == saml_session_id:
+                reqid = key
+        session_data: Optional[SamlAuthContext] = self._outstanding_requests.get(reqid)
+        if session_data:
+            logger.debug('session_data : %s', session_data)
+            del self._outstanding_requests[reqid]
+            return session_data.login
```

Comment:
The `saml_session_id` is assumed to exist and be found, but it may not(e.g. if I reload the `postacs.html`). If it is not, `reqid` will not be defined and there will be a `UnboundLocalError`(variable is not defined) at line 131.

```suggestion
        reqid = next((key for key, request in self._outstanding_requests.items() if request.saml_session_id == saml_session_id), None)
        if reqid is None:
            logger.debug('no outstanding request found for saml session %s', saml_session_id)
            return None
        # reqid was found, SamlAuthContext exists
        session_data = self._outstanding_requests[reqid]
        logger.debug('session_data : %s', session_data)
        del self._outstanding_requests[reqid]
        return session_data.login
```

Reply:
fixed